_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,416 @@
+      "topicId": "9fagJTJNTaa-FwT6umvvCA",
+      "partition": 0
+    }
+  },
+  "value": {
+    "version": 0,
+    "data": {
+      "snapshotEpoch": 2,
+      "stateEpoch": 0,
+      "leaderEpoch": 0,
+      "startOffset": 560,
+      "stateBatches": [
+        {
+          "firstOffset": 560,
+          "lastOffset": 560,
+          "deliveryState": 0,
+          "deliveryCount": 1
+        },
+        {
+          "firstOffset": 561,
+          "lastOffset": 564,
+          "deliveryState": 2,
+          "deliveryCount": 1
+        },
+        {
+          "firstOffset": 565,
+          "lastOffset": 565,
+          "deliveryState": 4,
+          "deliveryCount": 1
+        },
+        {
+          "firstOffset": 566,
+          "lastOffset": 569,
+          "deliveryState": 2,
+          "deliveryCount": 1
+        },
+        {
+          "firstOffset": 570,
+          "lastOffset": 570,
+          "deliveryState": 0,
+          "deliveryCount": 1
+        },
+      ]
+    }
+  }
+}
+```
+Subscribed partitions’ states are stored in batches as records get consumed and acknowledged in batches too. In the example above, you can see the batch of states, one of which is for offsets between 561 and 564. These offsets have the same delivery state and count therefore share the same state. In comparison, the offset 560 and 565 have different delivery states therefore listed in individual states.
+
+<i>Map of the states and enums:</i>
+
+| Enum | State |
+| :--: |:-------------|
+| 0    | Available    |
+| 2    | Acknowledged |
+| 4    | Archived     |
+
+Let’s take a look at the following example of a partition that a share group is subscribed to, which is called <b>Share Partition</b>:
+
+![Delivery state](/assets/images/posts/2025-06-24-kafka-queue-03.png)
+
+In this example, offset 2 is the start offset that a share group is consuming from the partition and offset 2 and 4 are currently acquired by a consumer of the group for the first time, therefore delivery count is 1.  However, the record at offset 3 is in <b>Available</b> state with delivery count of 2, which means a consumer of the group has attempted to deliver this record twice. This record will be retried until the maximum delivery count is reached. The record at offset 5 has been processed successfully and acknowledged and the record at offset 6 is the next available record to be acquired by a consumer of the share group, therefore it is the end offset for this share partition. 
+
+### Comparing share and consumer groups
+
+I have set up a single node cluster on my machine using Strimzi’s [Quickstart](https://strimzi.io/quickstarts/) and created a topic called <code>kafka-queue</code> with 2 partitions. 
+
+I also added the following 2 configurations to my Kafka CR to enable this feature, as it is not enabled by default since it is not production ready yet:
+```
+group.coordinator.rebalance.protocols: classic,consumer,share
+unstable.api.versions.enable: true
+```
```

Comment:
It would be good to share the Kafka CR with these values - for examle in simplified format, e.g.:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  # ...
spec:
  kafka:
    # ...
    config:
      group.coordinator.rebalance.protocols: classic,consumer,share
      unstable.api.versions.enable: true
```

Reply:
I think it might be nice to have the instructions here more directed at the reader, e.g. deploy the quickstart, then edit your Kafka CR to add these two settings...